tools/validation-generator/src/main/java/io/spine/tools/validate/MessageValidatorFactory.java

```diff
@@ -0,0 +1,201 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.tools.validate;
+
+import com.squareup.javapoet.AnnotationSpec;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.TypeSpec;
+import io.spine.annotation.Beta;
+import io.spine.code.gen.java.GeneratedBySpine;
+import io.spine.code.gen.java.NestedClassName;
+import io.spine.type.MessageType;
+import io.spine.validate.ConstraintViolation;
+import io.spine.validate.Constraints;
+import io.spine.validate.ValidationException;
+
+import javax.annotation.Generated;
+import java.util.Set;
+
+import static com.squareup.javapoet.ClassName.bestGuess;
+import static io.spine.tools.validate.ValidateMethod.immutableListOfViolations;
+import static java.lang.System.lineSeparator;
+import static javax.lang.model.element.Modifier.FINAL;
+import static javax.lang.model.element.Modifier.PRIVATE;
+import static javax.lang.model.element.Modifier.PUBLIC;
+import static javax.lang.model.element.Modifier.STATIC;
+
+/**
+ * A factory of message validation code.
+ */
+public final class MessageValidatorFactory {
+
+    private static final String VALIDATE_METHOD = "validate";
+    private static final String MESSAGE_VARIABLE = "msg";
+    private static final String VIOLATIONS = "constraintViolations";
+
+    private final MessageType type;
+    private final NestedClassName messageSimpleName;
+    private final String validatorSimpleName;
+
+    /**
+     * Creates a new {@code MessageValidatorFactory} for the given type.
+     *
+     * @param type
+     *         type of the message to validate
+     */
+    public MessageValidatorFactory(MessageType type) {
+        this.type = type;
+        this.messageSimpleName = NestedClassName.from(type.javaClassName());
+        this.validatorSimpleName = nameForValidator(type);
+    }
+
+    private static String nameForValidator(MessageType type) {
+        StringBuilder candidate = new StringBuilder("Validator");
+        while (nameClashes(type, candidate.toString())) {
+            candidate.append('$');
+        }
+        return candidate.toString();
+    }
+
+    private static boolean nameClashes(MessageType type, String name) {
+        boolean withOwnName = type.simpleJavaClassName()
+                                  .value()
+                                  .equals(name);
+        if (withOwnName) {
+            return true;
+        }
+        boolean withNestedName = type
+                .nestedDeclarations()
+                .stream()
+                .anyMatch(nested -> nested.simpleJavaClassName()
+                                          .value()
+                                          .equals(name));
+        return withNestedName;
+    }
+
+    /**
+     * Generates the {@code Validator} class for the associated message type.
```

Comment:
I don't think we have `Validator`s as a concept anymore. Maybe you could even reconsider the name of the `MessageValidatorFactory` itself.

Reply:
This is not the case. The Javadoc is up to date right now: the validation code is generated into a `private static final class Validator`.